ESPS/general/src/mcplay/mcplay.1
.\" Copyright (c) 1987 Entropic Speech, Inc.; All rights reserved
.\" @(#)mcplay.1	3.7 3/25/90 ESI
.ds ]W "\fI\s+4\ze\h'0.05'e\s-4\v'-0.4m'\fP\(*p\v'0.4m'\ Entropic Speech, Inc.
.TH MCPLAY 1\-ESPS 3/25/90
.SH NAME
mcplay - send portions of an ESPS file to the \s-1MASSCOMP\s+1 digital to analog converter.
play - send portions of an ESPS file to the digital to analog converter.
.SH SYNOPSIS
.B mcplay
[
.BI \-x " debug-level"
] [
.BI \-{psf} " range"
] [
.BI \-g " gain"
] [
.BI \-w " width"
] [
.BI \-r " nrepeat"
] [
.BI \-c " channel"
] [
.BI \-k " clock"
] [ 
.BI \-i 
] [
.BI \-h " hist-file"
] [
.BI \-b " shift-value"
] [
.BI \-C " clock device"
] [
.BI \-D " D/A device"
] [
.BI \-q 
]
.I file
[
.I more-files
]
.SH DESCRIPTION
.PP
.I Mcplay
sends a portion of one or more ESPS sampled data files
to a \s-1MASSCOMP\s+1 digital-to-analog converter.
A subrange of data within the files may be chosen; this subrange may be specified in
seconds, frames, or points. Furthermore, a repeat count, a scaling factor
(gain), shift factor, and an alternate channel number may be
specified.
.PP
If "-" is given for a filename, then the input is taken from standard
input.
.PP
The following options are supported:
.TP
.BI \-x " debug"
Only debug level 1 is defined in this version.   This causes several
messages to be printed as internal processing proceeds.   The default
is level 0, which causes no debug output.
.TP
.BI \-p " range"
Select a subrange of points to be played, using the format
.I start\-end
,
.I start:end 
or
.IR start:+count .
Either the start or the end may be omitted; the beginning or the end of the
file are used if no alternate is specified. 
.IP
If multiple files were specified, the same range from each file is played.
.TP
.BI \-s " range"
Select a subrange of speech, specified in seconds. For compatibility with
DSC LISTEN, the beginning of the file is 0.  The format for a subrange
is the same as the \fB\-f\fR option.  The arguments to the \fB-s\fR
option can be decimal numbers (eg. 1.5).
.TP
.BI \-f " range"
Select a subrange of speech, specified in frames. By default, a frame is
100 points, but this may be changed.
.TP
.BI \-w " width"
This switch defines the frame width. The width is ignored unless the
.B \-f
switch is also specified.
.TP
.BI \-r " count"
Repeat count. The selected speech is played
.I count
times. If multiple files were specified, the files are played in order, and
the order is repeated
.I count
times.
.TP
.BI \-c " channel"
Specifies an alternate channel number (default zero).
.TP
.BI \-k " clock"
Specifies an alternate clock number (default zero).
.TP
.BI \-g " gain"
Speech samples are multiplied by
.I gain
before playback.  Gain is a floating point value.
This overrides the automatic gain control described below.
.TP
.BI \-i 
If this flag is given and ESPS FEA_SD file generic header 
item \fImax_value\fR is zero, 
then any samples with a value greater than
the maximum of the D/A converter will be clipped.   If clipping
occurs, then an entry is made in a history file for each
sampled clipped.  This entry gives the sample number and the  value before
clipping.   At the end of the file a scale factor is given that will
prevent clipping. 
Use of this option, slows things down noticeably.
.TP
.BI \-h " hist-file"
Specifies an alternate history file.  The default is \fIplay.his\fR.
.TP
.BI \-b " shift-value"
If this option is given and \fIshift-value\fR is positive, then each
sample (after it is converted to integer format) is shifted to the right
by \fIshift-value\fR bits.   \fIShift-value\fR being negative causes a
left shift.   This option is useful to rescale data that has a higher
resolution than the D/A converter being used here.   For example, files
recorded by or synthesized from data recorded by a 16 bit A/D can be
shifted 4 bits to the right for correct output on a 12 bit D/A.    This
is preferred to using the \fB-g\fR option with a gain of .0625 because
the shift is faster than a floating multiply.   This difference is most
noticeable on large files.
.TP
.BI \-q
This option (q for quiet) suppresses all terminal output (except for
fatal error messages) from the program.  It is useful when play is being
called by another program.
.PP
The default clock device is /dev/dacp0/efclk0, and the default D/A
device is /dev/dacp0/daf0.   These defaults, might be different at your
site, however, since they can be changed by the ESPS installation
program.   The compiled in defaults can be replaced on the command line
with the \fB\-C\fR and the \fB\-D\fR options.
.PP
If the ESPS FEA_SD file generic header item \fImax_value\fR is greater than
the maximum value that the D/A converter can convert
(\s-1MAXDA\s+1), then each sample
is multiplied by \s-1MAXDA\s+1/\fImax_value\fR.
The effect is to scale the data so that the sample with the maximum
value is scaled to \s-1MAXDA\s+1.
.PP
Playback may be aborted by 
sending the terminal's interrupt character (normally control-C)
after playback has started. 
.PP
If any of the command line range options are given, they
override the \fIstart\fR and \fInan\fR common file values described
below.
.PP
This program is installed under the name \fImcplay\fR and \fIplay\fR.
Since this program is specific to \s-1MASSCOMP\s+1 systems, a different play program
is used on non-\s-1MASSCOMP\s+1 systems.   These system dependent play programs
all have a common user interface and are installed under their system
dependent name (eg. \fImcplay\fR, \fInetplay\fR, \fIsunplay\fR, etc) and
the name \fIplay\fR.
.PP
This version of the program only operates on single channel FEA_SD or SD
files.   Use \fIdemux\fR(1\-ESPS) to select channels from a
multi-channel FEA_SD file.   Only files containing non-complex fields
can be played.
.SH ESPS PARAMETERS
.PP
The parameter file is not read.  
.SH ESPS COMMON
.PP
ESPS Common processing may be disabled by setting the environment variable
USE_ESPS_COMMON to "off".  The default ESPS Common file is .espscom 
in the user's home directory.  This may be overidden by setting
the environment variable ESPSCOM to the desired path.  User feedback of
Common processing is determined by the environment variable ESPS_VERBOSE,
with 0 causing no feedback and increasing levels causing increasingly
detailed feedback.  If ESPS_VERBOSE is not defined, a default value of 3 is
assumed.
.PP
If Common file processing is enabled 
the following are read from the Common file if 
only one input file or no input file is given on the command line and
provided that standard input isn't used.  
.IP
.I "filename - string"
.IP
This is the name of the input file.  If no input file is 
specified on the command line, \fIfilename\fP is taken to be the
input file.  If an input file is specified on the command line, 
that input file name must match \fIfilename\fP or the other 
items (below) are not read from Common. 
.IP
.I "start - integer"
.IP
This is the starting point in the input file to begin output.  It 
is not read if the \fB\-p, -f,\fP or \fB-s\fR option is used.  
.sp
.I "nan - integer"
.IP
This is the number of points to output from the input file.  It is not
read if the \fB\-p, -f\fP or \fB-s\fR option is used.  A value of zero means the last
point in the file.
.sp
.I "gain - float"
.IP
This value is the same as the argument to the -g option described above.
This value is not read from Common if the \fB-g\fR or \fB-b\fR option is
used.
.sp
.I "shift - integer"
.IP
This value is the same as the argument to the -b option described above.
A gain value and a shift value cannot both be used.  If both are present
in the common file, then only the shift value is used.  
This value is not read from Common if the \fB-g\fR or \fB-b\fR option is
used.
.sp
.PP
Again, the values of \fIstart\fR and \fInan\fR are only used if the
input file on the command line is the same as \fIfilename\fP in the
common file, or if no input file was given on the command line.  If
\fIstart\fR and/or \fInan\fR are not given in the common file, or if
the common file can't be opened for reading, then \fIstart\fR
defaults to the beginning of the file and \fInan\fR defaults to the
number of points in the file.
.PP
The following items are written into the ESPS Common file:
.IP
.I "start - integer"
.IP
The starting point from the input file.
.sp
.I "nan - integer"
.IP
The number of points in the selected range.
.sp
.I "prog - string"
.IP
This is the name of the program (\fIplay\fP in this case).
.sp
.I "filename - string"
.IP
The name of the input file.  If multiple input files are processed, this
is the name of the first file.   
.sp
.I "gain - float"
.IP
This value is written if the \fB-g\fR option is used, or if \fIgain\fR was
read from Common.  
.sp
.I "shift - integer"
.IP
This value is written if the \fB-b\fR option is used, or if \fIshift\fR
was read from Common.
.PP
These items are not written to ESPS COMMON if the input file is <stdin>
or if there are more than one input files.
.SH DIAGNOSTICS
.PP
.I Mcplay
informs the user if the input file does not exist, or is not an ESPS sampled
data file, or if inconsistent options are used.
.PP
The program will timeout if the incorrect clock is connected.
.PP
If the starting point requested is greater than the last point in the
file, then a message is printed and the program exits with status 1.  If
the ending point requested is greater than the last point in the file,
it is reset to the last point, a warning is printed and processing
continues.
.SH SEE ALSO
mux(1\-ESPS), demux(1\-ESPS)
.SH BUGS
.SH AUTHOR
.PP
Alan Parker, based on the network play program by Joe Buck.


ESPS/general/src/mcplay/mcplay.c
/*

  This material contains proprietary software of Entropic Speech, Inc.   
  Any reproduction, distribution, or publication without the the prior	   
  written permission of Entropic Speech, Inc. is strictly prohibited.
  Any public distribution of copies of this work authorized in writing by
  Entropic Speech, Inc. must bear the notice			
 								
               "Copyright 1986,1987 Entropic Speech, Inc."
 				
 	
 									     
 Joseph T. Buck, Entropic Processing, Inc.				    
 Modified by Alan Parker to use DACP on Masscomp at WRL 2/13/86
 Modified for ESPS by Alan Parker, Entropic Speech, Inc.
 Modified by Joe Buck to play previous file while reading next one.

 This program is used to play a SD file on the Masscomp D/A. 
 							
						
 Syntax: mcplay [options] file		
   Options:			

	-x debug_level	

	-g gain	
		gain is a floating value. The file is multiplied by this    
		value to form the output data.
	-p range							  
		range is a range of points. Only those points are played 
		the first point is 1					
	-f range						
		range is a range of frames (default frame size = 100)	
		the first frame is 1				
	-w width					
		changes the frame size		
	-s range			
		range is a range of seconds. For LISTEN compatibility, the
		first second is zero					 
	-r repeats							
		number of repetitions				
	-c chan						
		channel number			
	
	-k clock
		D/A clock number
	-i
		Clip samples
	-h hist-file
		Specify alternate history file.
	-b
		bits to shift input data to form output data
		postive means right shift, negative means left shift

*/

#ifndef lint
	static char *sccs_id = "@(#)mcplay.c	3.13 11/8/90 ESI";
#endif


/* **************** check these defaults ********************** */

#ifndef DADEV
#define DADEV "/dev/dacp0/daf0"		/* default d to a device */
#endif
#ifndef CLKDEV
#define CLKDEV "/dev/dacp0/efclk"	/* default clock device */
#endif

#define CLKDEFAULT	0		/* default clock number */

#ifndef MAXDA
#define MAXDA 2048		/* max value of this D/A (12 bit) */
#endif

/* ************************************************************ */

#define NUM_BUFS 1
#define DUTY 50			/* duty cycle of clock */

#include <stdio.h>

/* include ESPS header files */
#include <esps/esps.h>			/* main header stuff */
#include <esps/fea.h>
#include <esps/feasd.h>			/* SD file stuff */
#include <esps/unix.h>			/* Unix function declarations */

#define SYNTAX USAGE \
("play [-[spf] range] [-w fsiz] [-c chan] [-k clock] [-r nrpt] [-g gain] [-i] [-h hist-file] [-x debug-level] file ...");

/* Globals */
int     debug_level = 0;

/* Externals */
void lrange_switch();
void frange_switch();
extern  int optind;
extern  char *optarg;
double	atof();
int     mropen(),  mrclktrig();
char    *getsym_s();
void    dac_error();
char	*eopen();
double  get_genhd_val();

main (argc, argv)
int argc;
char **argv;
{
    int     c, i;
    int     nrep = 1, channel = 0;
    char    *p_switch = NULL, *s_switch = NULL, *f_switch = NULL;
    FILE    *istrm, *hfile;
    char    *hist = "play.his";
    struct  header  *ih;
    long    start_p = 1, end_p, start_f, end_f;
    double   start_s, end_s;
    int     isf, np=0, fwidth = 100, gflag = 0;
    int	    bflag=0, bits, left, right;
    int     nptot = 0, clk = CLKDEFAULT, nofile=0;
    int     max = 0, clip = 0, iflag = 0;
    float   *dataf;
    short   *data = NULL;
    int     argind, devpn = -1, clkpn = -1, write_hfile=0;
    double  retwidth, retfreq, clk_width, gain=1;
    char    dadev[30], clkdev[30], *filename=NULL, *firstfile;
    int     filecount=0;
    int	    in_progress = 0, prev_np = 0, prev_isf = 0;
    int	    range_given = 0;
    int     qflag=0;	/* if 1 then suppress all normal messages */
    int     data_type; /* holds type of data in sampled data files*/


    (void)sprintf(dadev,DADEV); 
    (void)sprintf(clkdev,"%s%d",CLKDEV,clk);

    while ((c = getopt (argc, argv, "p:s:r:c:k:w:f:g:x:ih:b:C:D:q")) != EOF) {
	switch (c) {
	    case 'x':
		debug_level = atoi (optarg);
		break;
	    case 'g':
	        gflag++;
		gain = atof (optarg);
		break;
	    case 'c': 
		channel = atoi (optarg);
		break;
	    case 'r': 
		nrep = atoi (optarg);
		break;
	    case 'p': 
		range_given++;
		p_switch = optarg;
		break;
	    case 'f':
		range_given++;
		f_switch = optarg;
		break;
	    case 'w':
		fwidth = atoi (optarg);
		break;
	    case 's': 
		range_given++;
		s_switch = optarg;
		break;
	    case 'k':
		(void)sprintf(clkdev, "%s%s", CLKDEV, optarg);
		break;
	    case 'i':
		iflag++;
		break;
	    case 'h':
		hist = optarg;
		break;
	    case 'b':
		bflag++;
		bits = atoi(optarg);
		break;
	    case 'C':
		(void)strncpy(clkdev,optarg,sizeof clkdev);
		break;
	    case 'D':
		(void)strncpy(dadev,optarg,sizeof dadev);
		break;
	    case 'q':
		qflag=1;
		break;
	    default: 
		SYNTAX;
	}
    }

    if (bflag && gflag) {
	Fprintf(stderr,"play: can't use both -b and -g\n");
	exit(1);
    }

    if (optind >= argc)
      nofile = 1;
    else
      filename = argv[optind];

    (void)read_params(NULL,SC_CHECK_FILE,filename);
    if (nofile) {
      if(symtype("filename") == ST_UNDEF) {
	Fprintf(stderr,"play: no input file\n");
	SYNTAX;
      }
      filename = getsym_s("filename");
    }
    firstfile = filename;

    if ((hfile = fopen(hist,"w")) == NULL) {
	Fprintf(stderr,"play: warning can't open hist file.\n");
 	hfile = stderr;
    }

    if(debug_level) Fprintf(stderr,"play: clk: %s, d/a: %s\n",clkdev,dadev);
    argind = optind;

    while (nrep > 0) {
    optind = argind;
    while (optind < argc || nofile) {
	filecount++;
	if (!nofile) filename = argv[optind];
/* open the input file */
	filename = eopen("play", filename, "r", FT_FEA, FEA_SD, &ih, &istrm);

/* if input is a pipe (ndrec is -1) then assume 80000 points of data */
	if (ih->common.ndrec == -1)
	   ih->common.ndrec = 80000;

	end_p = ih -> common.ndrec;
	if((isf = *get_genhd_d("record_freq",ih)) <= 0) {
	   Fprintf(stderr,"play: %s sample frequency is <= zero; exiting\n",
			  filename);
	   exit(1);
	}
	start_s = (start_p - 1) / isf;
	end_s = (end_p + isf - 1) / isf;
	start_f = (start_p - 1) / fwidth + 1;
	end_f = (end_p + fwidth - 1) / fwidth;

	if(!range_given) {
	  if(symtype("start") != ST_UNDEF)
	     start_p = getsym_i("start");
	  if(symtype("nan") != ST_UNDEF)
	     end_p = start_p + getsym_i("nan") -  1;
	}

	if(!bflag && !gflag && symtype("shift") != ST_UNDEF) {
		bits = getsym_i("shift");
		Fprintf(stderr,
		 "play: using shift (-b option) value of %d from Common file.\n",bits);
		bflag++;
	}
	if(!gflag && !bflag && symtype("gain") != ST_UNDEF) {
		gain = getsym_d("gain");
		Fprintf(stderr,
		 "play: using gain (-g option) value of %lg from Common file.\n",gain);
		gflag++;
	}


        if (bflag)
          if (bits < 0) left = bits*-1;
	   else right = bits;

	if (p_switch)
	  lrange_switch (p_switch, &start_p, &end_p, 1);
	 else if (s_switch) {
	   frange_switch (s_switch, &start_s, &end_s);
	   start_p = start_s * isf + 1;
	   end_p = end_s * isf;
	   if (end_p <= start_p) end_p += isf;
	 }
	 else if (f_switch) {
	   lrange_switch (f_switch, &start_f, &end_f, 1);
	   start_p = (start_f - 1) * fwidth + 1;
	   end_p = end_f * fwidth;
	 }

	np = end_p - start_p + 1;

	if(start_p > ih->common.ndrec) {
	  Fprintf (stderr, "play: start point (%ld) > number of points (%ld)\n",
	  		start_p, ih -> common.ndrec);
	  exit(1);
	}

	if (start_p > end_p) {
	  Fprintf (stderr, "play: start point (%ld) > end point (%ld)\n",
		   start_p, end_p);
	  exit (1);
	}

	if (end_p > ih -> common.ndrec) {
	  Fprintf (stderr, "play: end point (%ld) > number of points (%ld)\n",
		   end_p, ih->common.ndrec);
	  Fprintf (stderr, "play: end point reset to last point.\n");
	  end_p = ih->common.ndrec;
	}

	if (!qflag)
	 Fprintf(stderr,"play: start point: %ld, length:%ld, filename: %s\n",
	 start_p,np,filename);
	
	skiprec (istrm, start_p - 1, size_rec (ih));

	clk_width = (double)((1.06e6/isf)*(DUTY/100.0));
	if(debug_level) Fprintf(stderr,"play: np: %ld\n",np);

	/*
	 * Check data type and allocate shorts if possible
	 */
	  data_type = get_sd_type(ih);
	  if(data_type == CHAR || data_type == SHORT){
	    data = (short *)malloc((unsigned)np * sizeof SHORT);
	    spsassert(data != NULL, "Cannot alloc SHORT buffer\n");
	  }
	  else{
	    dataf = (float *)malloc((unsigned)np * sizeof FLOAT);
	    spsassert(dataf != NULL, "Cannot alloc FLOAT buffer\n");
	  }


	if ((get_genhd_val("max_value",ih,0) == 0) && iflag) {
	    clip++; write_hfile=1;
	    Fprintf(hfile,
             "Samples greater than %d will be clipped to this value\n",
	      MAXDA);
	    Fprintf(hfile,"Sample\toriginal value\n");
	}

	if (!gflag && !bflag && (get_genhd_val("max_value",ih,0) > MAXDA)) {
	    gflag++;  write_hfile=1;
	    gain = (float)MAXDA/ *get_genhd_d("max_value",ih);
	    Fprintf(hfile,"hd.sd->max_value: %g\n",*get_genhd_d("max_value",ih));
	    Fprintf(hfile,"This D/A MAXDA is %d\n",MAXDA);
	    Fprintf(hfile,"Data scaled by %g\n",gain);
	    Fprintf(stderr,"play: Data scaled by %g\n",gain);
	}

        if(data_type == CHAR || data_type == SHORT){
	 if ((i = get_sd_recs(data,np,ih,istrm)) != np) {
	    if (i == 0) {
		Fprintf (stderr, "play: %s is empty!\n");
		exit (1);
	    }
	    else {
		Fprintf (stderr, 
			 "play: only %d points in %s\n", i, filename);
		np = i;
	    }
	 }
       }
       else{
	if ((i = get_sd_recf(dataf,np,ih,istrm)) != np) {
	    if (i == 0) {
		Fprintf (stderr, "play: %s is empty!\n");
		exit (1);
	    }
	    else {
		Fprintf (stderr, "play: only %d points in %s\n", i, filename);
		np = i;
	    }
	}
      }

	(void) fclose (istrm);
	if (clip) {
	  if (data_type != CHAR && data_type != SHORT){
            for (i=0; i<np; i++) {
                int j = abs((int)dataf[i]);
	        if (j > MAXDA) {
		   Fprintf(hfile, "%d\t%d\n", i+2,j);
		   if(j > max) max = j;
		   if(dataf[i] < 0) dataf[i] = -MAXDA;
		   else dataf[i] = MAXDA;
	        }
	    }
	  }
	  else{
            for (i=0; i<np; i++) {
                int j = abs((int)data[i]);
	        if (j > MAXDA) {
		   Fprintf(hfile, "%d\t%d\n", i+2,j);
		   if(j > max) max = j;
		   if(data[i] < 0) data[i] = -MAXDA;
		   else data[i] = MAXDA;
	        }
	    }
	  }
	}
/* Wait for previous play to complete.  Then allocate old array. */
	if (in_progress) {
	    wait_for_end(devpn,(int)(2000.0*prev_np/prev_isf),data);
	    Fprintf(stderr,"previous transfer complete\n");
	}
	prev_np = np;
	prev_isf = isf;
/* Re-open D/A */
	if(mropen(&devpn,dadev,0) == -1) {
	    Fprintf(stderr,"Can't open D/A %s\n",dadev);
	    exit(1);
        }
	if(mropen(&clkpn,clkdev,0) == -1) {
	    Fprintf(stderr,"Can't open clock %s\n",clkdev);
	    exit(1);
	}

	if(mrclk1(clkpn,0,(double)isf,&retfreq,0,clk_width,&retwidth,0) != 0)
    	 dac_error();
	if(mrclktrig(devpn,1,clkpn) != 0) dac_error();

	/* Allocate short buffer, if not already allocated */
        if(data_type != CHAR && data_type != SHORT){
	  if((data = (short *)malloc((unsigned)np*sizeof SHORT)) == NULL) {
	    Fprintf(stderr,"Cannot alloc %ld shorts\n",np);
	    exit(1);
	  }
	}

/* either, 1) multiply sample by gain factor (slow), 2) shift sample
   bits to right (faster), or 3) just copy sample (fastest)
 */
      if(data_type != CHAR && data_type != SHORT)
        {
	if (gflag) 
	    for (i=0; i<np; i++) data[i] = dataf[i]*gain;
	else if (bflag && left) 
	   for (i=0; i<np; i++) data[i] = (int)dataf[i] << left;
	else if (bflag && right)
	   for (i=0; i<np; i++) data[i] = (int)dataf[i] >> right;
	else
	   for (i=0; i<np; i++) data[i] = dataf[i];
        }
      else
	{
	if (gflag) 
	   for (i=0; i<np; i++) data[i] = data[i]*gain;
	else if (bflag && left) 
	   for (i=0; i<np; i++) data[i] = (int)data[i] << left;
	else if (bflag && right)
	   for (i=0; i<np; i++) data[i] = (int)data[i] >> right;
        }

	(void)mrdaxout(devpn,clkpn,clkpn,channel,1,1,np,data,0);

 	if(!qflag) Fprintf(stderr,"play: Transfer started\n");
        (void) free((char *)dataf);
	in_progress = 1;
        nptot += np;
	if(nofile) break;
        optind++;
    }
    nrep--;
    }
    wait_for_end(devpn,(np/isf)*2000,data);
    if(!qflag) Fprintf (stderr, "play: final transfer complete\n");

/* write common file, unless the input was from a pipe or there
   was more than one input file
*/
    if(filecount == 1 && strcmp(filename, "<stdin>") != 0) {
    	(void) putsym_i("start",(int)start_p);
    	(void) putsym_i("nan",(int)np);
    	(void) putsym_s("filename",firstfile);
    	(void) putsym_s("prog","play");
	if(gflag) (void)putsym_f("gain",gain);
	if(bflag) (void)putsym_i("shift",bits);
    }

    if(debug_level) Fprintf(stderr,"play: Total points output: %ld\n",nptot);
    if(clip && max) Fprintf(hfile,"To avoid clipping use a gain of %g.\n",
	(float)MAXDA/(float)abs(max));
    if(clip && max && (hfile != stderr)) 
	Fprintf(stderr,"play: To avoid clipping use a gain of %g.\n",
	(float)MAXDA/(float)abs(max));
    (void) fclose(hfile); 
    if(!write_hfile && (hfile != stderr)) (void)unlink(hist);
    return 0;
}

void dac_error()
{
    Fprintf(stderr,"Error on dacp call.\n");
    exit(1);
}

wait_for_end (devpn,arg_wt,data)
int devpn, arg_wt;
short *data;
{
    (void) mrevwt(devpn,0,arg_wt);
    if (data != NULL)
	(void) free((char *)data);
    if(mrclosall() != 0) {
	Fprintf (stderr, "Error from mrclosall call.\n");
	exit (1);
    }
}

